### ISSUE TITLE
Performance regression in process termination due to exception handling

### ISSUE THREAD

**Issue by guillep** [guillep/pharo-core#10354]
**Bug description**

We have observed a 1.25x performance degradation in process termination, affecting at least many tests. For example, all Calypso tests are impacted:

**To Reproduce**

```smalltalk
times := (1 to: 100) collect: [:rep |
	[ClyAllPackagesQueryTest package testSuite run ] timeToRun asMilliSeconds ].

times average asFloat.
times stdev
```

We have tracked down the regression to build 1402. In my machine, with a 100 in-process iterations the code above yields the following:

- Before 1402, time to run is 3015 +- 250 ms
- After 1402, time to run is 3765 +- 219 ms

We have observed that the issue comes from PR https://github.com/pharo-project/pharo/pull/9223.
And after some digging, the problem seems to come from line https://github.com/StevenCostiou/pharo/blob/a5b533d03e975dc9fae821df55f5a3c9affb11b1/src/Kernel/Context.class.st#L785

The problem being iterating all exception classes in the system to know if they are meant to be caught during process termination (and debugging).

We should have a smarter way to do it.

**Comment by guillep** [comment]
Same problem happens with RBMethodNodeTests in `AST-Core-Tests`.

**Comment by dionisiydk** [comment]
> And after some digging, the problem seems to come from line https://github.com/StevenCostiou/pharo/blob/a5b533d03e975dc9fae821df55f5a3c9affb11b1/src/Kernel/Context.class.st#L785
> 
> The problem being iterating all exception classes in the system to know if they are meant to be caught during process termination (and debugging).
> 
> We should have a smarter way to do it.

I pointed to the performance impact in this PR and suggested a different approach. But sadly the fix was not done after the integration

<blockquote><img src="https://opengraph.githubassets.com/5b04e7a39cfe1d607e7a37031157898ed9aa7be6e81d27c05b5356d1e95a5c50/StevenCostiou/pharo" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/StevenCostiou/pharo">pharo/Context.class.st at a5b533d03e975dc9fae821df55f5a3c9affb11b1 · StevenCostiou/pharo</a></strong></div><div>Pharo is a dynamic reflective pure object-oriented language supporting live programming inspired by Smalltalk. - pharo/Context.class.st at a5b533d03e975dc9fae821df55f5a3c9affb11b1 · StevenCostiou/p...</div></blockquote>

**Comment by Ducasse** [comment]
We know :) and we are lame to not have listen what you said.
Seriously. Thanks for your reviews.

**Comment by StevenCostiou** [comment]
But @dionisiydk you did the improvement of my bad implemetation here https://github.com/pharo-project/pharo/pull/9422 but it was reverted here https://github.com/tesonep/pharo/commit/fc3d754f31ab45d63423dd65325915b201001c93

I cannot find the reason of the revert, there might be something additionnal to fix.
Anyway, what you did here https://github.com/pharo-project/pharo/pull/9422  should improve things?

**Comment by dionisiydk** [comment]
Right. I forget about my PR and missed the revert. The revert PR #9445 does not explain what was broken

**Comment by guillep** [comment]
Hi @dionisiydk, we wanted to re-do/re-integrate now the changes in #9445 which seem good.
We are looking at why we reverted it, because we would like to prevent the same cause from appearing again.

Keep you posted.

**Comment by guillep** [comment]
This was related to process termination. When catching `Exception` there was a problem with the unwind of some processes.

Maybe the fixes to process termination we integrated solved that. We have to recheck.
